File: AOC_PY/days/day06.py
import os
from functools import reduce

INPUT_PATH = os.path.join(os.path.dirname(__file__), '../../Inputs', 'day06.txt')


def group_common_answers(group: str):
    return len(reduce(
        lambda acc, answer: acc.intersection(set(answer.strip())),
        group.split('\n'),
        set('abcdefghijklmnopqrstuvwxyz')
    ))


class Day06:

    def __init__(self):
        with open(INPUT_PATH, 'r') as infile:
            self.groups = [x.strip() for x in infile.read().split('\n\n')]

    def solve1(self):
        return reduce(lambda acc, group: acc + len(set(group.replace('\n', ''))), self.groups, 0)

    def solve2(self):
        return reduce(lambda acc, group: acc + group_common_answers(group), self.groups, 0)


def main():
    x = Day06()
    print(x.solve1())
    print(x.solve2())


if __name__ == '__main__':
    main()

File: AOC_CS/Days/Day06.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AOC_CS.Days
{
    class Day06
    {
        const string INPUT_FILE = "Inputs\\day06.txt";

        private string[] groups;

        public Day06()
        {
            groups = null;
        }

        private void ReadInput()
        {
            if (groups != null) return;
            groups = File.ReadAllText(INPUT_FILE).Split("\n\n");
            groups[groups.Length - 1] = groups.Last().Trim();
        }

        private int CountCommonAnswers(string group)
        {
            return group.Split('\n').Aggregate(
                new HashSet<char>("abcdefghijklmnopqrstuvwxyz"),
                (acc, val) => acc.Intersect(new HashSet<char>(val)).ToHashSet()
            ).Count;
        }

        public int Solve1()
        {
            ReadInput();
            return groups.Aggregate(0, (acc, val) => acc + new HashSet<char>(val.Replace("\n", "")).Count);
        }

        public int Solve2()
        {
            ReadInput();
            return groups.Aggregate(0, (acc, val) => acc + CountCommonAnswers(val));
        }
    }
}
